qt: Avoid potential -Wdeprecated-enum-enum-conversion warnings

hebasto opened the pull request:
This PR is related to bitcoin/bitcoin#24169. It adjusts code in order to avoid `-Wdeprecated-enum-enum-conversion` warnings instead of disabling them.

Could be tested with gcc 11.2.

hebasto committed:
qt: Use human-readable strings in preference to hard-coded integers

This is recommended by Qt docs.
See: https://doc.qt.io/qt-5/qkeysequence.html#details

Also this change avoids -Wdeprecated-enum-enum-conversion warnings.

hebasto committed:
qt: Avoid potential -Wdeprecated-enum-enum-conversion warning

fanquake commented on configure.ac:
```diff
@@ -449,6 +449,7 @@ if test "$CXXFLAGS_overridden" = "no"; then
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
   AX_CHECK_COMPILE_FLAG([-Wunreachable-code-loop-increment], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunreachable-code-loop-increment"], [], [$CXXFLAG_WERROR])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wimplicit-fallthrough"], [], [$CXXFLAG_WERROR])
+  AX_CHECK_COMPILE_FLAG([-Wdeprecated-enum-enum-conversion], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdeprecated-enum-enum-conversion"], [], [$CXXFLAG_WERROR])
```
Looks like his is enabled by default when targeting C++20 in GCC, and by default in Clang, so I don't think there's a need to add this explicitly.

MarcoFalke replied:
I presume it is not enabled by default in C++17, which is used to compile Bitcoin Core (and will be for the coming months/years)?

fanquake replied:
It's enabled by default in Clang, which should be good enough coverage for us.
In GCC, when targetting pre-C++20, it's enabled by `-Wenum-conversion`, which is part of `-Wextra`, which we already enable. Although in that case it may only be applied to C code.

hebasto replied:
> I presume it is not enabled by default in C++17

Correct.

MarcoFalke replied:
> It's enabled by default in Clang, which should be good enough coverage for us.

For me it is only enabled with c++20. Do you have steps to reproduce?

fanquake replied:
> Do you have steps to reproduce?

No. I assumed Clang might be enabling with any standard, given they [don't specify otherwise](https://clang.llvm.org/docs/DiagnosticsReference.html#wdeprecated-enum-enum-conversion), like GCC does. However, I'm still a NACK on adding this. It's already enabled by default when it's needed, it's only available on fairly "new" compilers, so having everyone building with older ones check for availability is pointless, and I'd prefer that our configure checks don't return to being a grab-bag of checks for random flags. In any case, given the only existence of warning producing code, are a few instances in the GUI, I'd say the likelihood of any warning producing code being introduced is a also very low.

hebasto replied:
> I'm still a NACK on adding this.

An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

fanquake replied:
> An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

I don't really think we should be adding warning flags to our build system for the sake of making testing PRs easier. If anyone wants to test this change, they can do `./configure CXXFLAGS="-Wdeprecated-enum-enum-conversion"` using a compiler that supports the option.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1075072189).

MarcoFalke commented on src/qt/bitcoingui.cpp:
```diff
@@ -251,28 +252,28 @@ void BitcoinGUI::createActions()
     overviewAction->setToolTip(overviewAction->statusTip());
     overviewAction->setCheckable(true);
-    overviewAction->setShortcut(QKeySequence(Qt::ALT + Qt::Key_1));
+    overviewAction->setShortcut(QKeySequence(tr("Alt+1")));
```
Isn't `tr` used to mark messages for translation? If yes, why is this needed here? If this is needed to convert to `QString`, why not call `QString{}` instead?

hebasto replied:
> Isn't `tr` used to mark messages for translation?

Yes, it is.

> If yes, why is this needed here? If this is needed to convert to `QString`, why not call `QString{}` instead?

No need for translation here. Updated in the recent push.

Thanks!

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -251,28 +252,28 @@ void BitcoinGUI::createActions()
     overviewAction->setToolTip(overviewAction->statusTip());
     overviewAction->setCheckable(true);
-    overviewAction->setShortcut(QKeySequence(Qt::ALT + Qt::Key_1));
+    overviewAction->setShortcut(QKeySequence(QString{"Alt+1"}));
```
0fea3846d0d6ec69dadd8bf043afd3c25971d07f

nit, here and below, use `QStringLiteral`, see https://doc.qt.io/qt-5/qstring.html#more-efficient-string-construction

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1074850209).

promag commented on src/qt/guiutil.cpp:
```diff
@@ -414,7 +415,7 @@ void bringToFront(QWidget* w)
 void handleCloseWindowShortcut(QWidget* w)
 {
-    QObject::connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), w), &QShortcut::activated, w, &QWidget::close);
+    QObject::connect(new QShortcut(QKeySequence(QShortcut::tr("Ctrl+W")), w), &QShortcut::activated, w, &QWidget::close);
```
0fea3846d0d6ec69dadd8bf043afd3c25971d07f

`QObject::tr()`

hebasto replied:
The initial idea was to put this translation into the context named "QShortcut", rather "QObject". But context name does not go from `src/qt/locale/bitcoin_en.ts` into the Transifex environment. So [accepting](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1074850209) your suggestion.

promag reviewed (commented):
Concept ACK Qt change.

hebasto commented:
Updated 38cffbb301e8d2ddf8dbc12afee99e76d8b54ebd -> 103b8dd8b89c772b7f6d097383e83d742b5ed55c ([pr24624.03](https://github.com/hebasto/bitcoin/commits/pr24624.03) -> [pr24624.04](https://github.com/hebasto/bitcoin/commits/pr24624.04), [diff](https://github.com/hebasto/bitcoin/compare/pr24624.03..pr24624.04)):

- addressed @promag's comments

MarcoFalke commented:
Can you explain why any of this needs translation? Is it displayed anywhere to the user?

hebasto commented:
> Can you explain why any of this needs translation? Is it displayed anywhere to the user?

![Screenshot from 2022-03-22 09-16-01](https://user-images.githubusercontent.com/32963518/159436822-65602247-4c09-44d8-acd3-e00368991f3a.png)

hebasto commented:
Updated 103b8dd8b89c772b7f6d097383e83d742b5ed55c -> acd98adaf1d83b71eda235c49d41a92f30c16313 ([pr24624.04](https://github.com/hebasto/bitcoin/commits/pr24624.04) -> [pr24624.05](https://github.com/hebasto/bitcoin/commits/pr24624.05), [diff](https://github.com/hebasto/bitcoin/compare/pr24624.04..pr24624.05)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/24624#discussion_r832060992)

fanquake reviewed (approved):
untested ACK acd98adaf1d83b71eda235c49d41a92f30c16313 - thanks.

MarcoFalke commented:
Approach ACK acd98adaf1d83b71eda235c49d41a92f30c16313

For testing locally, #24169 will need to be merged.

fanquake commented:
> For testing locally, https://github.com/bitcoin/bitcoin/pull/24169 will need to be merged.

Why? You can just do `./configure CXXFLAGS="-Wdeprecated-enum-enum-conversion"` with a supporting compiler.

promag reviewed (commented):
Code review ACK acd98adaf1d83b71eda235c49d41a92f30c16313.
